### PR TITLE
Improve pitch/roll_comp, logical_intervals, state_intervals, get_ofp_states, get_telem_table

### LINKED ISSUE #247: pitch_comp roll_comp computed msids fail
For machines or users without internet / maude access, the pitch_comp / roll_comp computed msids will fail.  This seems largely benign, but is there a way to determine if the query is fully contained within the 'cxc' source data before making calls to maude?

For the below I just removed my .netrc to confirm this fails.

```
In [3]: dat = fetch.Msid('pitch_comp', "2023:040", "2023:050")
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /proj/sot/ska3/test/lib/python3.10/site-packages/ska_helpers/utils.py:123, in LazyDict.__getitem__(self, item)
    122 try:
--> 123     return super().__getitem__(item)
    124 except KeyError:

KeyError: 'version'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 dat = fetch.Msid('pitch_comp', "2023:040", "2023:050")

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/fetch.py:1844, in Msid.__init__(self, msid, start, stop, filter_bad, stat)
   1843 def __init__(self, msid, start=LAUNCH_DATE, stop=None, filter_bad=True, stat=None):
-> 1844     super(Msid, self).__init__(
   1845         msid, start=start, stop=stop, filter_bad=filter_bad, stat=stat
   1846     )

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/fetch.py:572, in MSID.__init__(self, msid, start, stop, filter_bad, stat)
    570 # Get the times, values, bad values mask from the HDF5 files archive
    571 if intervals is None:
--> 572     self._get_data()
    573 else:
    574     self._get_data_over_intervals(intervals)

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/fetch.py:637, in MSID._get_data(self)
    635 comp_cls = ComputedMsid.get_matching_comp_cls(self.msid)
    636 if comp_cls:
--> 637     self._get_comp_data(comp_cls)
    638     return
    640 # Avoid stomping on caller's filetype 'ft' values with _cache_ft()

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/fetch.py:703, in MSID._get_comp_data(self, comp_cls)
    700 logger.info(f"Getting computed values for {self.msid}")
    702 # Do computation.  This returns a dict of MSID attribute values.
--> 703 attrs = comp_cls(self.units["system"])(
    704     self.tstart, self.tstop, self.msid, self.stat
    705 )
    707 # Allow upstream class to be a bit sloppy on times and include samples
    708 # outside the time range.  This can happen with classes that inherit
    709 # from DerivedParameter.
    710 ok = (attrs["times"] >= self.tstart) & (attrs["times"] <= self.tstop)

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/derived/comps.py:250, in ComputedMsid.__call__(self, tstart, tstop, msid, interval)
    244 match_args = [
    245     arg.lower() if isinstance(arg, str) else arg for arg in match.groups()
    246 ]
    248 if interval is None:
    249     # Call the actual user-supplied work method to compute the MSID values
--> 250     msid_attrs = self.get_msid_attrs(tstart, tstop, msid.lower(), match_args)
    252     for attr in ("vals", "bads", "times", "unit"):
    253         if attr not in msid_attrs:

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/derived/comps.py:683, in Comp_Pitch_Roll_OBC_Safe.get_msid_attrs(self, tstart, tstop, msid, msid_args)
    680 # Whether we are computing "pitch" or "roll", parsed from MSID name
    681 pitch_roll: str = msid_args[0]
--> 683 ofp_states = get_ofp_states(stop.date, days=(stop - start).jd)
    685 tlms = []
    686 for ofp_state in ofp_states:

File /proj/sot/ska3/test/lib/python3.10/site-packages/kadi/commands/utils.py:122, in get_ofp_states(stop, days)
    119 from cheta.utils import logical_intervals
    121 msid = "conlofp"
--> 122 tlm = get_telem_values([msid], stop, days)
    123 states_list = []
    124 for state_code in np.unique(tlm[msid]):

File /proj/sot/ska3/test/lib/python3.10/site-packages/kadi/commands/utils.py:85, in get_telem_values(msids, stop, days)
     82 logger.info(f"Fetching telemetry for {msids} between {start.date} and {stop.date}")
     84 with fetch.data_source("cxc", "maude allow_subset=False"):
---> 85     msidset = fetch.MSIDset(msids, start.date, stop.date)
     87 # Use the first MSID as the primary one to set the time base
     88 msid0 = msidset[msids[0]]

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/fetch.py:1554, in MSIDset.__init__(self, msids, start, stop, filter_bad, stat)
   1552 new_msids = []
   1553 for msid in msids:
-> 1554     new_msids.extend(msid_glob(msid)[0])
   1555 for msid in new_msids:
   1556     if intervals is None:

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/fetch.py:401, in msid_glob(msid)
    399 sources = data_source.sources(include_test=False)
    400 for source in sources:
--> 401     ms, MS = _msid_glob(msid, source)
    402     msids.update((m, None) for m in ms)
    403     MSIDS.update((m, None) for m in MS)

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/fetch.py:426, in _msid_glob(msid, source)
    415 def _msid_glob(msid, source):
    416     """Get the archive MSIDs matching ``msid``.
    417 
    418     The function returns a tuple of (msids, MSIDs) where ``msids`` is a list of
   (...)
    424     :returns: tuple (msids, MSIDs)
    425     """
--> 426     source_msids = data_source.get_msids(source)
    428     # MSID is the upper-case version of the MSID name that is actually used in
    429     # backend queries. ``msid`` is the user-supplied version.
    430     MSID = msid.upper()

File /proj/sot/ska3/test/lib/python3.10/site-packages/cheta/fetch.py:155, in _DataSource.get_msids(cls, source)
    152 elif source == "maude":
    153     import maude
--> 155     out = list(maude.MSIDS.keys())
    156 else:
    157     raise ValueError('source must be "cxc" or "msid"')

File /proj/sot/ska3/test/lib/python3.10/site-packages/ska_helpers/utils.py:11, in _lazy_load_wrap.<locals>.wrapper(self, *args, **kwargs)
      9 @functools.wraps(unbound_method)
     10 def wrapper(self, *args, **kwargs):
---> 11     self._load()
     12     return unbound_method(self, *args, **kwargs)

File /proj/sot/ska3/test/lib/python3.10/site-packages/ska_helpers/utils.py:113, in LazyDict._load(self)
    111 def _load(self):
    112     if not self._loaded:
--> 113         self.update(self._load_func(*self._args, **self._kwargs))
    114         self._loaded = True
    116         # Delete these so serialization always works (pickling a func can fail)

File /proj/sot/ska3/test/lib/python3.10/site-packages/maude/maude.py:108, in load_msids()
    107 def load_msids():
--> 108     msids = {msid: index for index, msid in MSID_FROM_INDEX.items()}
    110     # Establish alias of DP_<MSID> as a synonym for CALC_<MSID>
    111     for msid in list(msids):

File /proj/sot/ska3/test/lib/python3.10/site-packages/ska_helpers/utils.py:11, in _lazy_load_wrap.<locals>.wrapper(self, *args, **kwargs)
      9 @functools.wraps(unbound_method)
     10 def wrapper(self, *args, **kwargs):
---> 11     self._load()
     12     return unbound_method(self, *args, **kwargs)

File /proj/sot/ska3/test/lib/python3.10/site-packages/ska_helpers/utils.py:113, in LazyDict._load(self)
    111 def _load(self):
    112     if not self._loaded:
--> 113         self.update(self._load_func(*self._args, **self._kwargs))
    114         self._loaded = True
    116         # Delete these so serialization always works (pickling a func can fail)

File /proj/sot/ska3/test/lib/python3.10/site-packages/maude/maude.py:95, in load_msid_index()
     94 def load_msid_index():
---> 95     with open(get_data_filename('MSIDIDX', 'txt'), 'r') as fh:
     96         lines = fh.read().splitlines()
     98     msid_index = {}

File /proj/sot/ska3/test/lib/python3.10/site-packages/maude/maude.py:71, in get_data_filename(basename, suffix)
     70 def get_data_filename(basename, suffix):
---> 71     version = MAUDE_SYSINFO['version']
     72     for ver in ('-' + version, ''):
     73         filename = os.path.join(os.path.dirname(__file__), basename + ver + '.' + suffix)

File /proj/sot/ska3/test/lib/python3.10/site-packages/ska_helpers/utils.py:125, in LazyDict.__getitem__(self, item)
    123     return super().__getitem__(item)
    124 except KeyError:
--> 125     self._load()
    126     return super().__getitem__(item)

File /proj/sot/ska3/test/lib/python3.10/site-packages/ska_helpers/utils.py:113, in LazyDict._load(self)
    111 def _load(self):
    112     if not self._loaded:
--> 113         self.update(self._load_func(*self._args, **self._kwargs))
    114         self._loaded = True
    116         # Delete these so serialization always works (pickling a func can fail)

File /proj/sot/ska3/test/lib/python3.10/site-packages/maude/maude.py:129, in get_maude_sysinfo(user, password)
    120 """
    121 Get system info about the MAUDE server.
    122 
   (...)
    126 :returns: dict of system information
    127 """
    128 if user is None or password is None:
--> 129     user, password = get_user_password()
    130 url = MAUDE_URL + 'sysinfo.json'
    131 logger.debug('Getting URL {}'.format(url))

File /proj/sot/ska3/test/lib/python3.10/site-packages/maude/maude.py:411, in get_user_password()
    406 else:
    407     # Try reaching MAUDE server with a light query for system version info
    408     # to see if we can get away without authentication (e.g. client is on
    409     # a network with no-auth access).
    410     url = MAUDE_URL + 'sysinfo.json'
--> 411     r = get_request(url, timeout=5)
    412     if r.status_code == 200:
    413         user, password = None, None

TypeError: get_request() missing 2 required positional arguments: 'user' and 'password'

```

### PR DESCRIPTION
**Pull request by taldcroft**
## Description

This is a hodge-podge of related improvements to utilities in cheta. The initial driver was a problem in `[aca] Kadi commands state validation: ALERT` on Apr 1 ending with:
```
 File "/proj/sot/ska3/flight/lib/python3.10/site-packages/cheta/utils.py", line 334, in state_intervals
    raise ValueError("Filtered data length must be at least 2")
ValueError: Filtered data length must be at least 2
```

Generally speaking the original implementation of `pitch_comp` and `roll_comp` was not robust. It depended on `get_ofp_states` in `kadi.utils` which was also not robust. So this turned into an effort to revisit this code and make it work under most circumstances. In particular dealing with short query intervals was a focus.

Since the code hinges on `state_intervals` and `logical_intervals`, those got pulled into the effort.

This is used by https://github.com/sot/kadi/pull/285.

### Changes
- Fix `pitch_comp` and `roll_comp` to work for short query intervals like normal MSIDs, namely to return values if the interval covers a time where the relevant primary telemetry are available. Since this computation uses telemetry sampled from 1.025 sec up to 5 minutes (ephemeris), this is not completely trivial. In any case it should always return a result, even if that result has no data. Added new tests of this.
- Add `start` and `stop` args to `logical_intervals` and `state_intervals`. This is useful when you want the table to states to exactly span the requested interval. This is different from the default behavior where the states start and end times are defined by the input times.
- For `logical_intervals` and `state_intervals` provide special handling for cases with input data of length 0 and 1. Previously you would get some non-obvious exceptions.
- Add the `get_telem_table` function which was migrated and generalized somewhat from the kadi validate code.

TODO:
- [x] Add more tests for `state_intervals`

Fixes #247

### Unit tests

- [ ] No unit tests
- [x] Mac
- [ ] Linux
- [ ] Windows

Independent check of unit tests by Jean
- [x] Linux (fido)

### Functional tests

No functional testing.

### CODE REVIEW

**Comment by jeanconn** [comment]
Will this also close #247 ?

**Comment by taldcroft** [comment]
> Will this also close https://github.com/sot/cheta/issues/247 ?

Yes! I've updated the description.

**Comment by jeanconn** [comment]
This has an open TODO of "Add more tests for state_intervals".  Should that get cut or should I review in advance of that (if it is ongoing work)?

File: cheta/derived/comps.py [hunk 1]
```diff
@@ -554,8 +554,28 @@ def get_msid_attrs(self, tstart, tstop, msid, msid_args):
 
 
 @functools.lru_cache(maxsize=1)
-def get_roll_pitch_tlm(start, stop):
```
Comment (jeanconn) [inline cheta/derived/comps.py:557]: Looks like this is replaced by better methods, but removing it looks like an interface impact.

File: cheta/derived/comps.py [hunk 2]
```diff
@@ -554,8 +554,28 @@ def get_msid_attrs(self, tstart, tstop, msid, msid_args):
...
+        bads |= (dat[msid].vals < x0) | (dat[msid].vals > x1)
+    if np.any(bads):
+        dat = dat[~bads]
+
+    return dat
+
+
+@functools.lru_cache(maxsize=1)
+def get_roll_pitch_tlm_safe(start, stop):
+    """Get telemetry values to compute pitch and roll in safe mode"""
```
Comment (jeanconn) [inline cheta/derived/comps.py:577]: Might help our code search to note that these are CSS data msids

**Comment by jeanconn** [comment]
I'm seeing test failures on kady
```
============================================= short test summary info ==============================================
FAILED cheta/tests/test_intervals.py::test_logical_intervals_start_stop_1 - AssertionError: assert ['duration ts...   3.5   4.5'] == ['duration ts...   3.5   6.0']
FAILED cheta/tests/test_intervals.py::test_logical_intervals_start_stop_2 - AssertionError: assert ['duration ts...   3.5   4.5'] == ['duration ts...   3.5   6.0']
=================================== 2 failed, 76 deselected, 9 warnings in 3.02s ===================================
ska3-jeanconn-kady> 
```
Does this need to be tested with another change?

File: cheta/derived/comps.py (hunk 1 above)
Comment (taldcroft) [reply cheta/derived/comps.py:557]: I did that and it is now checked.

**Comment by taldcroft** [comment]
About the TODO, I did that and it is now checked.

File: cheta/derived/comps.py (hunk 1 above)
Comment (taldcroft) [reply cheta/derived/comps.py:557]: This is an internal method used only in this module and it is not in the public API `__all__`.

File: cheta/derived/comps.py (hunk 2 above)
Comment (taldcroft) [inline cheta/derived/comps.py:577]:
```suggestion
    """Get telemetry values to compute pitch and roll in safe mode.
    
    This uses the OBC-computed sun position in the solar array frame from CSS data.
    """
```

Comment (taldcroft) [reply cheta/derived/comps.py]: Fixed.

**Comment by taldcroft** [comment]
About the test failure, I'm seeing that now on Mac in my branch. No idea when that broke and why I didn't see it before. Hence why we have that checkbox for someone else to run the tests at the end.

**Comment by taldcroft** [comment]
@jeanconn - I think I have resolved all the issues with a code fix and new tests and a rebase.

**Pull request merged**